### ISSUE TITLE
Merge spam and reverse firers in favor of "reversal types"

### ISSUE THREAD

**Issue by JDGrimes** [WordPoints/hooks-api#79]
I think that maybe it would make sense for us to merge the spam firer and reverse firer. Instead, we'd have just the reverse firer, but it would recognize different kinds of reversals, of which `spam` would be one.

One motivation for this is the fact that when a reactor listens to both spam and reverse fires, it would have to jump through hoops to make sure that it didn't result in duplicate reversals. In other words, both the spam and reverse firers would be triggered by a spam reversal (I think). It would be simpler to let the reactor check what type of reversal was taking place. That way it could handle spam different from other types of reversals if it chose, and it wouldn't have to check if a reverse fire was going to have a duplicate spam fire accompanying it (because there would be just one reverse fire).

Another thing to consider is the possibility that we'll want to introduce more types of reverse fires in the future. This would be a nightmare if we introduce a firer for each one. But if we just have a new reversal type, it would be possible for the reactor to keep handling it the way it normally would any other reversal.

Also, these two firers are pretty much duplicates of each other, so it would by DRYer to merge them.

**Comment by JDGrimes** [comment]
Possibly we'll want to do something similar in introducing different types of regular fires as well.

**Comment by JDGrimes** [comment]
## How to determine reversal type?

The big question is, how do we determine what type of reversal a reversal is? Where should that information come from? Currently what determines whether an reversal is considered regular or spam is how the actions are registered for the event. If the action is registered as a spam action for that event, then the reverse would be performed by the spam firer, and the therefore it would be a spam reversal. In other words, the type of a reverse was supplied by the dev when composing the actions for the event.

I'm wonder now though if possibly we want to change that entirely, and instead determine the type of reverse from the primary event arg. After all, it is the status which the primary event arg is being transitioned to that determines the type of the event. To do this, we'd have to invent some sort of entity status API, of course.

Note that we can't supply this information at the action level, because actions can be reused for multiple events. It could be the reversal action for one event but the firing action for another. So the action itself doesn't tell us what sort of fire is taking place, that can only be determined from the args.

**Comment by JDGrimes** [comment]
## Behavior differences of the multi-reverser vs unified firer

One thing to consider is that keeping these separate would have made it possible for a spam reversal to occur even after a regular reversal occurred. In other words, when the spam and reverse firers are separate and can fire independently, it is possible that each could occur independently from the other. We mentioned above the possibility that they could both occur together, but this would actually depend upon the actions which were registered for the event. It would also be possible for the reverse to occur, then the spam, or for the spam alone to occur, or for the reverse to occur later, separate from the spam. Basically, the current arrangement allows for anything to happen.

What we must decide then is whether all of these options should really be available, and whether they are possible with a merged API.

First, let's consider the case when the spam comes without the reverse, and the reverse only comes independently at some later point if at all. That should never happen, because a reverse is a reverse, no matter what type. The `spam` reversal is a sub-type of the `reverse` reversal. Which is why we said that they should be merged.

But I suppose that an alternative would be to reject the `reverse` arch-type, and instead elevate `spam` to being a first-class citizen alongside `draft`, `trash`, etc. The problem with that, as pointed out above, is that it would be almost impossible to implement new types of reversals in the future, unless they used a unified API (which is essentially what we are proposing).

So some form of restriction needs to be supplied in the API, to make sure that every spam also triggers a regular reverse. Our proposal of "merging" the two firers would be one means of accomplishing that, by triggering a reverse of the spam type.

Now let's consider the case when the reverse occurs on its own, and the spam does not occur until later, independently, if at all.

Is that something that we want to remain possible in the new API?

Actually, I'm not sure that we've ever given much thought to the fact that this is possible. And I should note that the emphasis is on _possible_. It is possible to do this, but it is actually not necessary for this to be possible, the API doesn't require it to work that way—it entirely depends upon how the actions are implemented and registered for the event. So this might be possible for one event but not for another.

That only leads to confusion, so again there is a need to do something here regardless of whether merging the firers is the means of achieving it. We need to either make this a requirement or make it impossible (merging the firers and doing #68 would accomplish the latter).

But which behavior is correct?

The fact is that continuing to listen for spam after a reverse has occurred would be an anomaly in an API that is mostly not stateful. Unless and until we begin listening for every time a post is edited to check if more points should be awarded (see #69), I think this would be the wrong direction to take. So for the initial version of the API at least, we need to unify around not listening for different kinds of reversals after the initial reversal (though of course it could be added later, even within a unified firer as proposed—in fact, it would be easier to change things like this when using a unified approach instead of many different firers).

In other words, none of the ways that current multi-reverser approach can operate are correct:
- It shouldn't be possible for a spam to occur without a reversal.
- It shouldn't be possible for a spam to occur after a reversal.
- And as we pointed out at the start, it shouldn't be possible for a spam to occur along with a reversal.

In other words, the current approach is flawed, and does not work right at all. The unified firer would not only solve all of these, but would make it easier for us to innovate in the future.

**Comment by JDGrimes** [comment]
## Do we need multiple reversal types?

I think before we go too far here we should ask the most fundamental question: do we really need to support multiple types of reversals? Or should we just forget about differentiating spam from everything else within our unified reverse firer?

### Why we have it

First let's consider why we created multiple reversal types when building the reversal API.

The fact is that reversals have always suffered a sort of crises of purpose. Should reversals be nothing more than a spam detection mechanism, or should it be more than that? What really is the purpose of reversals. This goes back to the very start of the points hooks API.

As described in #2, most recently the points hooks API has been moving toward making reversals purely a spam detection mechanism. That is the direction we've continued to take reversals from the start in this new API:

> ###Reversals' Future
> 
> The current trajectory of reversals in WordPoints is that they are an automatic background task that helps stop users who try to game the system. On the principle of "decisions, not options", and that spam just shouldn't matter, this seems like a good path to take. It does raise a question though: Might reversals have a broader application in the future, beyond just stopping cheaters?
> 
> The answer to this question is no. This is what the reversal API is all about. That is its sole purpose. Any other use for reversals will either have to fit within the API, or use its own.

This led us to ultimately split the spam and reversal APIs, using only the spam API for the points reactor. The reversal API was going to be unused in the initial version, but was included so that other reactors could take advantage of it if they wanted to listen for all kinds of reversals (the badges reactor was in mind). In the end we decided that this was too big of a change to make as part of the new API, and that we'd have to postpone the day of reckoning on how points and reversals should really behave. So currently we are planning to have the points reactor begin listening to all reversals, not just spam, although this has not been implemented in the code yet. We haven't really discussed whether we'll continue to handle spam differently from other reversals, but I guess there's been no reason (until now perhaps) to change it.

The important thing to keep in mind is that the reversal API is no longer a slave of the points reactor. It is still used by that (for now), but the reversal API's behavior shouldn't be determined solely by what the points reactor wants. The points reactor is perfectly capable of handling reversals in a special way if it wants to. But reversals are a general thing for all reactors to use, and so should work in the most general way that common sense will allow.

That said, currently the points reactor does expect to handle spam differently from other types of reversals. At the very least, it is likely that we'll want it to do that in the future, though I suppose we could always introduce a reaction types API then.

In short, the main reason that we have different reversal types is because the points reactor would like to handle spam reversals differently from other reversals.

### Problems with reversal types

One thing that bugs me is the fact that it is the event itself that must be considered spam, not the entity, yet we determine that based on whether the entity was spam or not. This makes some sense, but it is not a guaranteed thing for every event. We are marking the hits as spam, and the entity itself doesn't have to be marked as spam for the event to have been spam. So even if we introduce an entity status API, it won't be sufficient for determining when an event has been spammed.

I suppose that we should really ask the question, should it be the event that is spam or the entity itself? In other words, is this way of thinking of it correct? I guess to answer that we'd need to explore what the difference between them is, and why it matters.

If the entity itself must be given a spam status, it would restrict which events could have the spam reversal type. An event in which the status of the entity did not change, and so the event was repeatable, would not be "spammable". Or, the reversal could be flagged as a spam reversal I suppose, if the entity happen to have the spam status, but that wouldn't always be correct. 

It would also make it impossible for an event which could only occur when an entity became spam to have the spam reversal type.

Basically, what this demonstrates is that even introducing a full-fledged entity status API wouldn't be all we need to identify reversal types.

In addition to that (or possibly instead of it?) we would need to make event reverse type detection a part some other piece of the API, like the action API or of the event args API.

We can't make it a part of the action API very well, because each action object can be used for multiple different events. That leaves us with the event args API only. But that doesn't make a great deal of sense either, because the arg is just a means of encapsulating the entity, and as we said above, the entity itself doesn't necessarily determine what type of reversal is taking place. That is also largely dependent on the action. Also, the arg is used both for the fire and the reverse, yet we're not going to be using a corresponding fire type API, just a reverse type API. So it doesn't really seem logical to make the revers type detection API a part of the args API.

### Conclusion

The only way that we could really do this would be to make it a part of the event API via the event objects themselves. We'd have to introduce a method that would try to determine what type of reverse was taking place. But then it would probably need access to both the event args and the event action in order to be able to determine that in every case.

Of course, we could just continue to have the developer supply the event type manually in some way. That would still leave the possibility for us to introduce an API to help automatically determine this in a future version. It might even be more forward compatible than just doing nothing. At the moment I'm leaning toward doing just that. If we keep the event reversal types in the API at all, that is.

**Comment by JDGrimes** [comment]
The above doesn't really give us a clear answer. The reason that we have reversal types is because the points reactor wanted to handle them differently. Specifically, it wanted to be able to treat regular reverses differently than when users are gaming the system. While it is still possible that it will want to do that in the future, right now I'm leaning toward just treating all reversals the same. I think we need to get a better idea of what users expect before we mess with that.

For that reason, I'm leaning toward not including reversal types in the initial API. We can always take the time to develop it later when it is actually needed and we have a better idea what we want from it and how it should work.

The only thing that would hold me back is if there was an eminent need for it for other reactors. However, I can't thing of any reason that there would be—after all we introduced this specifically for the points reactor and haven't been planning to use it for any others yet. 

So for the moment I'm not going to pursue reversal types in this API.

**Comment by JDGrimes** [comment]
As a side note, the stateful arg detection API (see #64) is related to this, since it allows the detection of args that change state during the event, though it doesn't tell what state they change to. However, that is mostly for the initial fire of the event, not so much for reversals. In fact, I suppose that reversals should perhaps be allowed to happen even when the initial arg doesn't have its state changed, because it is the event itself being marked as spam, not the entity. Anyway, it is also probable that we might change how the primary arg of an event is identified before release anyway.
